coding/src/analysis.py

```diff
@@ -0,0 +1,423 @@
+            self.results['score_' + str(k)] = self.top_k_match(probs, labels, k)
+    
+    def group_by_columns(self, columns, df=None):
+        '''
+        Groups df by columns and returns a dataframe of the aggregated scores.
+        Arguments:
+            columns (list): list of columns to group by
+            df (pandas dataframe): dataframe to group. Default is self.results
+        '''
+        if df is None:
+            df = self.results
+        df = df.groupby(columns).agg({'score': 'mean'})
+        return df
+    
+    def average_predictions(self, additional_columns=None):
+        '''
+        Average predictions for a given title and category to look at an "ensembled" prediction.
+        Arguments:
+            additional_columns (list): list of additional columns to split over. Default is None.
+        '''
+        columns = ['title', 'category']
+        if additional_columns is not None:
+            columns += additional_columns
+        # aggregate columns = self.categories
+        agg_dict = {cat: 'mean' for cat in self.categories}
+        df = self.results.groupby(columns).agg(agg_dict)
+        df = df.reset_index()
+
+        df['guess'] = df[self.categories].idxmax(axis=1)
+        df['score'] = 1 * (df['guess'] == df['category'])
+        return df
+    
+    def plot(self, df=None, x_variable='n_exemplars', y_variable='score', split_by=[], color_by='', average=False, save_path=''):
+        '''
+        Plots the results of x_variable vs. y_variable.
+        Arguments:
+            x_variable (str): column to plot on x-axis
+            y_variable (str): column to plot on y-axis
+            split_by (list): list of columns to split by. Default is []
+            color_by (str): column to color by. Must be included in split_by. Default is None.
+            save_path (str): path to save figure. Default is ''
+        '''
+        if df is None:
+            df = self.results
+
+        # check if split_by is a list, if not, make it a list
+        if not isinstance(split_by, list):
+            split_by = [split_by]
+        
+        if color_by:
+            # check if color_by is in split_by, if not, add it
+            if color_by not in split_by:
+                split_by.append(color_by)
+            unique = df[color_by].unique()
+            colors = sns.color_palette("Set1", len(unique))
+            color_dict = {cat: color for cat, color in zip(unique, colors)}
+            # df['color'] = df[color_by].map(color_dict)
+            # split_by.append('color')
+        
+        # add x_variable to split_by
+        if x_variable not in split_by:
+            split_by.append(x_variable)
+
+        df_agg = df.groupby(split_by).agg({y_variable: 'mean'})
+        df_agg = df_agg.reset_index()
+
+        # iterate through unique tuples of columns, excluding x_variable and y_variable
+        other_columns = list(set(df_agg.columns) - set([x_variable, y_variable]))
```

Comment:
When are there gonna be more than one of each column or x_variable or y_variable?

Reply:
This is just to make the function more robust - if the user listed x_variable or y_variable as a split_by, then it won't throw an error. Makes no difference functionally, just is less strict about the arguments.